WordPress/gutenberg
# Move the block manager to block editor package

Part of https://github.com/WordPress/gutenberg/issues/41004

## What?
Moves the block manager component to the block-editor package.

## Why?
It will allow the block manager to be more easily implement in other editors (like the site editor).

Now that `hiddenBlockTypes` are stored in the preferences store (since #39132), this change is easier to make.

## How?
It's worth discussing the changes that require most attention from reviewers first. To make it easier for other editors to implement the block manager, I've made some changes to the block editor settings.

Previously, the block manager worked by modifying the `allowedBlocks` setting (removing any hidden block types). It also set another `defaultAllowedBlocks` editor setting, which represented the allowed blocks before any modification.

I've changed this. It seems much simpler to use `hiddenBlockTypes` as an editor setting rather than `defaultAllowedBlockTypes` and make `canInsertBlockTypeUnmemoized` check the `hiddenBlockTypes` list, so that's the change I've made.

I'm open to other option, including changing it back. An alternative could also be for the block editor package to pull in `hiddenBlockTypes` straight from the preferences store instead of using a setting, but it'd have to somehow know which `scope` to use.

The rest of the changes are pretty straightforward:
- The block manager component is moved over
- I've switched it over from using `withSelect` to `useSelect`
- It accepts a new `scope` prop to indicate which preferences it uses.
- It uses a new internal `useHiddenBlockTypes` hook to interact with the preferences store (which could be worth exporting and then the post editor's selectors and actions could be deprecated).

## Testing Instructions
There shouldn't be any change in functionality in the post editor. To test:
1. Open the editor preferences modal
2. Go to the block section
3. Toggle the visibility of various blocks
4. Close the modal and open the inserter

Expected: The blocks that were hidden are not visible in the inserter.
